### ISSUE TITLE
Tracking issue for RFC 2388, "Reserve `try` for `try { .. }` block expressions"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#50412]
This is a tracking issue for the RFC "Reserve `try` for `try { .. }` block expressions" (rust-lang/rfcs#2388).

This issue only tracks the keyword reservation in edition 2018, and renaming `do catch { .. }` to `try { .. }`, for the main issue, see https://github.com/rust-lang/rust/issues/31436.

**Steps:**

- [ ] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?
                                      cc @Manishearth, @scottmcm )
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])
     - note: stabilization of `try { .. }` is tracked via #31436.

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

None.

**Comment by kennytm** [comment]
> Ensure that `try { .. } $ident { .. }` does not parse  (so that we have options for `try { .. } catch { .. }`).

Huh this is not mentioned in the RFC at all.

(Currently `do catch { stuff } $ident { ... }` won't *typecheck* because `()` is not `Try`)

**Comment by est31** [comment]
Where is the part that the try macro should still be allowed? What happened to that?

**Comment by nikomatsakis** [comment]
As I mentioned to @est3I believe that we ought to be able to support `try!` as a kind of "special case" in the parser.

**Comment by Centril** [comment]
@nikomatsakis
> Seems worth adding as an unresolved question

Done :)

**Comment by Centril** [comment]
cc @Manishearth on `try!` in edition 2018.

**Comment by scottmcm** [comment]
@Centril I think work here is currently blocked on https://github.com/rust-lang/rust/pull/50307

**Comment by est31** [comment]
[Very interesting cpp proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf ) just as this RFC wants to bury `try!(expr)`: they want to add `try expr` expressions.

A big point for this RFC was familiarity with other languages. Now as finally other languages are becoming familiar *to us* we get rid of those now familiar components and adopt their legacy idioms. This isn't really good.

So why not be consistent with Cpp here and adopt their `try expr` expressions? Then instead of `try!(expr)` and `expr?` you could write `try expr`, just like that Cpp proposal says.

**Comment by scottmcm** [comment]
@est31 That paper is consistent with this RFC's use of `try`, as its try expressions are still marking _areas_ in which exceptions are possible, not _every call_ which can throw.  From 4.5.2 (p38):
```c++
try return “xyzzy”s + “plover”;             // ok, covers both “”s and +
return try “xyzzy”s + “plover”;             // same
```
If it were like `?`, then it'd need to be this other example from the section, but it's explicitly not needed
```c++
return try ((try “xyzzy”s) + “plover”);     // ok, but redundant
```
Overall, the `try expr` feels more like the [Automatic “?” inside try {…} blocks](https://internals.rust-lang.org/t/automatic-inside-try-blocks/7507?u=scottmcm) IRLO thread.

**Comment by alexreg** [comment]
I'm yet to see a good explanation of the semantics of all this `try` business... it looks nasty to me, but what is it?

**Comment by Nemo157** [comment]
@alexreg see "`catch { ... }` expressions" on the [other tracking issue linked in the OP](https://github.com/rust-lang/rust/issues/31436), this issue is just about renaming those from their current implementation of `do catch { ... }` to `try { ... }`.

**Comment by alexreg** [comment]
@Nemo157 Okay, that makes the semantics of `try { ... }` clear, but I still don’t get what `try { ... } catch { ... }` should do...

**Comment by alexreg** [comment]
@cramertj Oh good. Just scoping of the `?` operator in other words? Changing its exit behaviour.

**Comment by glmdgrielson** [comment]
Wait, why does this need to be an unconditional keyword? I'm a bit uncomfortable about this since this means that code somebody just copied may break when pasted into a new file for no obvious reason. Why can't we handle it like `union`? What's the difference?

**Comment by kennytm** [comment]
@glmdgrielson See the discussion in #31436. In short, 

```rust
struct try { a: Option<u32> }

let a = Some(1);
let b = try { a }; // is this a `try` expression or a struct literal?
```

**Comment by scottmcm** [comment]
@glmdgrielson The difference with `union` is that it's always followed by an ident, so it can be disambiguated with one token of lookahead, since `ident ident` is never a valid expression.  And that's helped by the fact that `union` is an item, so in most places it's obvious, since `.union` or `(union` or similar can't be items and `pub union` (or just `union` outside a `fn`) can't be an expression.

**Comment by scottmcm** [comment]
Added this to the edition milestone for tracking purposes, since this issue is all about restricting in the edition.  (For extra clarity: stabilization of the feature is https://github.com/rust-lang/rust/issues/31436, which is **not** part of the edition.)

**Comment by SimonSapin** [comment]
I was mislead by the title of this RFC. I thought it was only about reserving the keyword to have another option available. The title should have indicated that it also makes a "normative" change to syntax introduced in a previously-accepted RFC.

**Comment by pnkfelix** [comment]
visted for T-compiler triage. It seems like progress here has ... stalled? Does someone own the remaining steps here, given that it is on the 2018 Release Milestone?

**Comment by pnkfelix** [comment]
also, tagging as P-high

**Comment by Centril** [comment]
@pnkfelix I believe that everything edition sensitive has been done wrt. keyword reservation and such.

**Comment by Centril** [comment]
@scottmcm yeah I think we can close it... should we perhaps open a new tracking issue for `try { ... }` ?

**Comment by viper6277** [comment]
> As I mentioned to @est3I believe that we ought to be able to support `try!` as a kind of "special case" in the parser. Seems worth adding as an unresolved question, at least (this would basically be a kind of contextual keyword). I'd still however prefer that in Rust 2018 `try!` the macro is deprecated (and warns you, suggesting a rewrite to use `?` -- though perhaps not a "rustfix-enabled" rewrite). At minimum I suppose we can rewrite to `r#try!(...)`.

Newbie here....So ...how do I replace  ?

```rust
r#try!(m.add(py, "test", py_fn!(py, test())));
```

**Comment by cramertj** [comment]
`m.add(py, "test", py_fn!(py, test()))?;`

**Comment by viper6277** [comment]
> `m.add(py, "test", py_fn!(py, test()))?;`

Thanks @cramertj .... run into these syntax issues all the time with rust, but I still love what the language is about, ... I've become a huge rust fan and plan on sticking with it.
